language/move-lang/src/cfgir/liveness/mod.rs

```diff
@@ -0,0 +1,356 @@
+}
+
+fn command(state: &mut LivenessState, sp!(_, cmd_): &Command) {
+    use Command_ as C;
+    match cmd_ {
+        C::Assign(ls, e) => {
+            lvalues(state, ls);
+            exp(state, e);
+        }
+        C::Mutate(el, er) => {
+            exp(state, er);
+            exp(state, el)
+        }
+        C::Return(e) | C::Abort(e) | C::IgnoreAndPop { exp: e, .. } | C::JumpIf { cond: e, .. } => {
+            exp(state, e)
+        }
+
+        C::Jump(_) => (),
+    }
+}
+
+fn lvalues(state: &mut LivenessState, ls: &[LValue]) {
+    ls.iter().for_each(|l| lvalue(state, l))
+}
+
+fn lvalue(state: &mut LivenessState, sp!(_, l_): &LValue) {
+    use LValue_ as L;
+    match l_ {
+        L::Ignore => (),
+        L::Var(v, _) => {
+            state.live.remove(v);
+            state.assigned.insert(v.clone());
+        }
+        L::Unpack(_, _, fields) => fields.iter().for_each(|(_, l)| lvalue(state, l)),
+    }
+}
+
+fn exp(state: &mut LivenessState, parent_e: &Exp) {
+    use UnannotatedExp_ as E;
+    match &parent_e.exp.value {
+        E::Unit | E::Value(_) | E::UnresolvedError => (),
+
+        E::BorrowLocal(_, var) | E::Copy { var, .. } => use_local(state, var),
+
+        E::Move { var, .. } => {
+            use_local(state, var);
+            state.moved.insert(var.clone());
+        }
+
+        E::ModuleCall(mcall) => exp(state, &mcall.arguments),
+        E::Builtin(_, e)
+        | E::Freeze(e)
+        | E::Dereference(e)
+        | E::UnaryExp(_, e)
+        | E::Borrow(_, e, _) => exp(state, e),
+
+        E::BinopExp(e1, _, e2) => {
+            exp(state, e1);
+            exp(state, e2)
+        }
+
+        E::Pack(_, _, fields) => fields.iter().for_each(|(_, _, e)| exp(state, e)),
+
+        E::ExpList(es) => es.iter().for_each(|item| exp_list_item(state, item)),
+    }
+}
+
+fn exp_list_item(state: &mut LivenessState, item: &ExpListItem) {
+    match item {
+        ExpListItem::Single(e, _) | ExpListItem::Splat(_, e, _) => exp(state, e),
+    }
+}
+
+fn use_local(state: &mut LivenessState, local: &Var) {
+    state.live.insert(local.clone());
+}
+
+//**************************************************************************************************
+// Refinement
+//**************************************************************************************************
+
+/// This refinement releases dead reference values by adding a move + pop. In other words, if a
+/// reference `r` is dead, it will insert `_ = move r` after the last usage
+///
+/// There are two cases where it will consider a reference to be "dead"
+/// (1) The reference was live at the previous command, but is not. However! No pop is inserted if
+///     the last usage `move`d the value
+/// (2) The previous command assigned to a reference holding local, but the local is not used
+///
+/// Examples for case (1)
+///
+/// ```move
+/// t0() {
+///   let x = 0;
+///   let x_ref = &mut x;
+///   *x_ref = 0;
+///   // inserted _ = move x_ref;
```

Comment:
Wouldn't it be better to convert the use of `x_ref` above into a move?

Reply:
That's a different, forth coming analysis :)